### PR TITLE
Redo stm32h7 SPI state machine

### PR DESCRIPTION
**Pull request by cbiffle**
My last round of fixes left things in a sort of mostly-working state, but I didn't like that very much, so I revisited the state machine.

I have also included a fix for SPI4/5/6 on STM32H7, which have shallower FIFOs.

### CODE REVIEW

File: drv/stm32h7-spi-server/src/main.rs [hunk 1]
```diff
@@ -345,7 +345,7 @@ impl ServerImpl {
         // The BufReader/Writer types manage position tracking for us.
 
         // Wrap a buffer reader/writer onto whichever borrows actually exist.
-        const BUFSIZ: usize = 32;
+        const BUFSIZ: usize = 16;
```
Comment (mkeeter) [inline drv/stm32h7-spi-server/src/main.rs:348]: Why did this change?  A comment on how this value is chosen would be helpful, although maybe it's obvious if I knew more about `LeaseBufReader`'s behavior.

**Review by mkeeter — APPROVED** [review]
Tested on the `vsc7448` branch, and it comes up just fine at the default SPI speed (`DIV64`).

If I increase the speed, I see crashes in the `vsc7448` task, but it looks like it's because the VSC7448 chip is unhappy: it's returning `0x88888888`, which is a sign that we're reading too fast.  This isn't an issue with the SPI driver, so I think we're good to merge.

File: drv/stm32h7-spi-server/src/main.rs (hunk 1 above)
Comment (cbiffle) [reply drv/stm32h7-spi-server/src/main.rs:348]: Honestly I was just messing around, neither value shows significantly improved performance at this point so I went for the "use less stack" option. I can add a comment.

Comment (cbiffle) [reply drv/stm32h7-spi-server/src/main.rs:351]: I have added a comment!

**Comment by cbiffle** [comment]
> Tested on the `vsc7448` branch, and it comes up just fine at the default SPI speed (`DIV64`).
> 
> If I increase the speed, I see crashes in the `vsc7448` task, but it looks like it's because the VSC7448 chip is unhappy: it's returning `0x88888888`, which is a sign that we're reading too fast. This isn't an issue with the SPI driver, so I think we're good to merge.

@mkeeter that's 3.125 MHz, which would be a surprisingly low top speed. Are you running through dupont wires?

**Comment by mkeeter** [comment]
@cbiffle The `0x88888888` issue isn't the SPI's problem; the VSC7448 needs a certain amount of time to pull data from its internal registers, so you have to configure padding bits if your SPI clock is particularly fast.  There's a long comment about it [here](https://github.com/oxidecomputer/hubris/pull/391/files#diff-835ba993bc30ceb6719833b4c6ec19dd29aff9bf2079a5dab8f2f91dc7f300b2R75-R92).

If I configure the system to run with more padding bytes, then it works at `DIV32`, but the `spi2_driver` still locks up at `DIV16`.  It's no longer doing an interrupt storm, but it's stuck here:

```
system time = 1481
ID TASK                 GEN PRI STATE
 6 spi2_driver            1   2 notif: bit0
   |
   +--->  0x200022b8 0x08029b08 userlib::sys_recv_stub
   |                 @ /hubris//sys/userlib/src/lib.rs:331
   |      0x20002368 0x08028d64 userlib::sys_recv
   |                 @ /hubris/sys/userlib/src/lib.rs:279
   |      0x20002368 0x08028d64 userlib::sys_recv_closed
   |                 @ /hubris/sys/userlib/src/lib.rs:260
   |      0x20002368 0x08028d74 drv_stm32h7_spi_server::ServerImpl::ready_writey
   |                 @ /hubris//drv/stm32h7-spi-server/src/main.rs:254
   |      0x200023e8 0x080286b8 drv_stm32h7_spi_server::<impl idol_runtime::Server<drv_stm32h7_spi_server::SpiOperation> for (core::marker::PhantomData<drv_stm32h7_spi_server::SpiOperation>,&mut S)>::handle
   |                 @ /hubris/target/thumbv7em-none-eabihf/release/build/drv-stm32h7-spi-server-877b944e122afbd5/out/server_stub.rs:197
   |      0x200023e8 0x080285dc idol_runtime::dispatch
   |                 @ /git/idolatry-1ebf1c2fd2f30300/478f0b6/runtime/src/lib.rs:169
   |      0x200023e8 0x08028896 main
   |                 @ /hubris//drv/stm32h7-spi-server/src/main.rs:63
   |
   |
   +--->   R0 = 0x0802a52c   R1 = 0x00000000   R2 = 0x00000001   R3 = 0x20002320
           R4 = 0x0802a52c   R5 = 0x00000000   R6 = 0x00000001   R7 = 0x8000ffff
           R8 = 0x200023c0   R9 = 0x00000000  R10 = 0x00000001  R11 = 0x00000001
          R12 = 0x20002430   SP = 0x20002298   LR = 0x08028d75   PC = 0x08029b08
```

Looking at the ringbuf, it looks like it wrote data successfully, but hasn't received anything:

```
humility: ring buffer drv_stm32h7_spi_server::__RINGBUF in spi2_driver:
 NDX LINE      GEN    COUNT PAYLOAD
   0  303        1        1 Start(write, (0x7, 0x0))
   1  417        1        1 Tx(0x80)
   2  417        1        2 Tx(0x0)
   3  417        1        4 Tx(0x81)
   4  481        1        2 WaitISR(0x70012)
```

I also don't see anything suspicious in the SPI registers:
```
# humility readmem -w 0x40003800 64
humility: attached via OpenOCD
                   \/        4        8        c
0x40003800 | 00001201 00000007 30070007 a4400000 | ...........0..@.
0x40003810 | 00000009 00070012 00000000 00000000 | ................
0x40003820 | 00000000 00000000 00000000 00000000 | ................
0x40003830 | 00000000 00000000 00000000 00000000 | ................
```

**Pull request merged**